README.md
[![Build
Status](https://ci.wbhatti.org/buildStatus/icon?job=authnzerver)](https://ci.wbhatti.org/job/authnzerver) [![Documentation Status](https://readthedocs.org/projects/authnzerver/badge/?version=latest)](https://authnzerver.readthedocs.io/en/latest/?badge=latest)

This is a small server meant to help add authentication (authn) and
authorization (authz) to other HTTP servers. It's built using
[Tornado](http://www.tornadoweb.org), [SQLAlchemy](https://www.sqlalchemy.org/),
[cryptography](https://cryptography.io),
[argon2-cffi](https://argon2-cffi.readthedocs.io/en/stable/),
[python-diskcache](http://www.grantjenks.com/docs/diskcache/), and
[uvloop](https://github.com/MagicStack/uvloop).

I wrote it to help with the login/logout/signup flows for the
[LCC-Server](https://github.com/waqasbhatti/lcc-server) and extracted much of
the code from there. It builds on the auth bits there and is eventually meant to
replace them. It can do the following things:

- Handle user sign-ups, logins, logouts, and locks/unlocks.
- Handle user email verification, password changes, forgotten password
  processes, and editing user properties.
- Handle API key issuance and verification.
- Handle access and rate-limit checks for arbitrary schemes of user roles,
  permissions, and target items. There is a [default
  scheme](https://github.com/waqasbhatti/authnzerver/blob/master/authnzerver/default-permissions-model.json)
  of permissions and user roles, originally from the LCC-Server where this code
  was extracted from. A custom permissions policy can be specified as JSON.

See
[TODO.md](https://github.com/waqasbhatti/authnzerver/blob/master/docs/TODO.md) for
features that are planned for the future. See
[CHANGELOG.md](https://github.com/waqasbhatti/authnzerver/blob/master/CHANGELOG.md)
for a version history.

Authnzerver talks to a frontend server over HTTP. Communications are secured
with symmetric encryption using the [cryptography](https://cryptography.io)
package's [Fernet scheme](https://cryptography.io/en/latest/fernet/), so you'll
need a pre-shared key that both Authnzerver and your frontend server know.

See [the HTTP API docs](https://authnzerver.readthedocs.io/en/latest/api.html)
for details on how to call Authnzerver from a frontend service.

More docs will be available at Authnzerver's (in-progress) [Read The
Docs](https://authnzerver.readthedocs.io/en/latest/) site.


## Installation

Authnzerver is [available at PyPI](https://pypi.org/project/authnzerver/), but
is very much a work in progress at the moment. Maybe hold off on installing it
until we've reached v0.2 (beta).

With that said, it can be installed (preferably in a virtualenv) using `pip`:

```bash
(venv) $ pip install authnzerver

# use pip install authnzerver --pre for unstable releases
```


## Running the server

There is a single executable that will be in your `$PATH` if you have a
virtualenv activated and the package installed: `authnzrv`.

`authnzrv --help` will list all the options available:

```
--authdb         An SQLAlchemy database URL to indicate where
                 the local authentication DB is. This should
                 be in the form discussed at: https://docs.sq
                 lalchemy.org/en/latest/core/engines.html#dat
                 abase-urls
--autosetup      If this is True, will automatically generate
                 an SQLite authentication database in the
                 basedir if there isn't one present and the
                 value of the authdb option is also None.
                 (default False)
--basedir        The base directory containing secret files
                 and the auth DB. (default os.getcwd())
--cachedir       Path to the cache directory to be used.
                 (default /tmp/authnzerver-cache)
--confvars       Path to the file containing the configuration
                 variables needed by the server and how to load
                 them. (default /path/to/authnzerver/
                 authnzerver/confvars.py)
--debugmode      If 1, will enable an /echo endpoint for
                 debugging purposes. (default False)
--emailpass      The password to use for login to the email
                 server.
--emailport      The SMTP port of the email server to use.
                 (default 25)
--emailsender    The account name and email address that the
                 authnzerver will send from. (default
                 Authnzerver <authnzerver@localhost>)
--emailserver    The address of the email server to use.
                 (default localhost)
--emailuser      The username to use for login to the email
                 server. (default getpass.getuser())
--envfile        Path to a file containing environ variables
                 for testing/development.
--listen         Bind to this address and serve content.
                 (default 127.0.0.1)
--permissions    The JSON file containing the permissions
                 model the server will enforce. (default
                 install-dir/authnzerver/authnzerver/
                 default-permissions-model.json)
--piisalt        A random value used as a salt when SHA256
                 hashing personally identifiable information
                 (PII), such as user IDs and session tokens,
                 etc. for authnzerver logs.
--port           Run the server on this TCP port. (default
                 13431)
--secret         The shared secret key used to secure
                 communications between authnzerver and any
                 frontend servers.
--sessionexpiry  This sets the session-expiry time in days.
                 (default 30)
--workers        The number of background workers to use when
                 processing requests. (default 4)
```

There's an example systemd `.service` file available in the `deploy` directory
to run this server automatically on startup.


## Configuring the server

Use the following environmental variables to configure the server. Defaults are
noted below where appropriate.

```
# listen address, port settings, and workers
AUTHNZERVER_PORT=13141
AUTHNZERVER_LISTEN=127.0.0.1
AUTHNZERVER_WORKERS=4

# cache and base directory locations
AUTHNZERVER_CACHEDIR=/tmp/authnzerver-cache
AUTHNZERVER_BASEDIR=directory where the server is started

# secret token, PII salt, and authentication DB URL
AUTHNZERVER_SECRET=
AUTHNZERVER_PIISALT=
AUTHNZERVER_AUTHDB=

# session expiry time in days
AUTHNZERVER_SESSIONEXPIRY=30

# permissions model JSON
AUTHNZERVER_PERMISSIONS=path/to/default-permissions-model.json

# email settings for sending emails to users
AUTHNZERVER_EMAILSENDER=Authnzerver <authnzerver@localhost>
AUTHNZERVER_EMAILSERVER=localhost
AUTHNZERVER_EMAILPORT=25
AUTHNZERVER_EMAILUSER=user running the authnzrv executable
AUTHNZERVER_EMAILPASS=''
```

You can also provide all of these at once using an environment file. This is not
recommended for production but is useful for development. If you go this route,
use the `--envfile` option to point to an appropriate environment file.

At a minimum, you must provide:

- a random pre-shared secret key as an environmental variable:
  `AUTHNZERVER_SECRETKEY`.
- a random salt value for hashing personally identifiable information in the
  authnzerver logs as an environmental variable: `AUTHNZERVER_PIISALT`.
- an SQLAlchemy [database
  URL](https://docs.sqlalchemy.org/en/latest/core/engines.html#database-urls)
  indicating where the server should store authentication information as an
  environmental variable: `AUTHNZERVER_AUTHDB`.

If none of these are provided, and the command-line option
`--autosetup=True`, the server will prompt you for admin credentials during
start up, generate the pre-shared secret key and random salt, and initialize an
authentication database at the SQLAlchemy URL you provide. Autogenerated
defaults for these values can be used by hitting Enter at all the prompts.


## License

Authnzerver is provided under the MIT License. See the LICENSE file for details.


CHANGELOG.md
# authnzerver v0.1.3 - 2020-03-30

## New stuff

- Email addresses for user sign-up are now checked against a list of disposable
  email provider domains.
- Configuration variables can be now retrieved from a variety of sources: from
  an environment variable, from a file, or from an HTTP request. The method of
  retrieval is now customizable by editing the `confvars.py` file copied to the
  server's base directory after `authnzrv --autosetup` is run.
- The `default-permissions-model.json` file is now copied to the server's base
  directory when `authnzrv --autosetup` is run for easier customization of the
  permissions policy.

## Changes

- The server now checks if the user's password needs to be rehashed when they
  sign-in. This should enable automatic upgrades of the Argon2 parameters as
  required.

## Fixes

- If a configuration variable is None after retrieval from a source, the server
  now halts start-up instead of continuing with that value.


# authnzerver v0.1.2 - 2020-03-27

## New stuff

- Added Sphinx generated autodocs for modules:
  https://authnzerver.readthedocs.io/en/latest/.


# authnzerver v0.1.1 - 2020-03-27

## Fixes

- Various minor documentation fixes.


# authnzerver v0.1.0 - 2020-03-26

This is the initial release.


authnzerver/confvars.py
# -*- coding: utf-8 -*-
# main.py - Waqas Bhatti (wbhatti@astro.princeton.edu) - Aug 2018
# License: MIT - see the LICENSE file for the full text.

'''Contains the configuration variables that define how the server operates.

The CONF dict in this file describes how to load these variables from the
environment or command-line options.

You can change this file as needed. It will be copied over to the authnzerver's
base directory when ``authnzrv --autosetup`` is run and you can tell authnzerver
to use it like so: ``authnzrv --confvars /path/to/basedir/confvars.py``.

You MUST NOT store any actual secrets in this file; just define how to get to
them.

For example, look at the ``secret`` dict entry below in CONF::

    'secret':{
        'env':'%s_SECRET' % ENVPREFIX,
        'cmdline':'secret',
        'type':str,
        'default':None,
        'help':('The shared secret key used to secure '
                'communications between authnzerver and any frontend servers.'),
        'readable_from_file':'string',
        'postprocess_value':None,
    }

This means the server will look at an environmental variable called
``AUTHNZERVER_SECRET``, falling back to the value provided in the ``--secret``
command line option. The ``readable_from_file`` key tells the server how to
handle the value it retrieved from either of these two sources.

To indicate that the retrieved value is to be used directly, set
``"readable_from_file" = False``.

To indicate that the retrieved value can either be: (i) used directly or, (ii)
may be a path to a file and the actual value of the ``secret`` item is a string
to be read from that file, set ``"readable_from_file" = "string"``.

To indicate that the retrieved value is a URL and the authnzerver must fetch the
actual secret from this URL, set::

    "readable_from_file" = ("http",
                            {'method':'get',
                             'headers':{header dict},
                             'data':{param dict},
                             'timeout':5.0},
                             'string')

Finally, you can also tell the server to fetch a JSON and pick out a key in the
JSON. See the docstring for :py:func:`authnzerver.confload.get_conf_item` for
more details on the various ways to retrieve the actual item pointed to by the
config variable key.

To make this example more concrete, if the authnzerver ``secret`` was stored as
a `GCP Secrets Manager
<https://cloud.google.com/secret-manager/docs/creating-and-accessing-secrets#access_a_secret_version>`_
item, you'd set some environmental variables like so::

    GCP_SECMAN_URL=https://secretmanager.googleapis.com/v1/projects/abcproj/secrets/abc/versions/z:access
    GCP_AUTH_TOKEN=some-secret-token

Then change the ``secret`` dict item in CONF dict below to::

    'secret':{
        'env':'GCP_SECMAN_URL',
        'cmdline':'secret',
        'type':str,
        'default':None,
        'help':('The shared secret key used to secure '
                'communications between authnzerver and any frontend servers.'),
        'readable_from_file':see below,
        'postprocess_value':'custom_decode.py::custom_b64decode',
    }

The ``readable_from_file`` key would be set to something like::

    "readable_from_file" = ("http",
                            {"method":"get",
                             "headers":{"Authorization":"Bearer [[GCP_AUTH_TOKEN]]",
                                        "Content-Type":"application/json",
                                        "x-goog-user-project": "abcproj"},
                             "data":None,
                             "timeout":5.0},
                            'json',
                            "payload.data")

This would then load the authnzerver ``secret`` directly from the Secrets
Manager.

Notice that we used a path to a Python module and function for the
``postprocess_value`` key. This is because GCP's Secrets Manager base-64 encodes
the data you put into it and we need to post-process the value we get back from
the stored item's URL. This module looks like::

    import base64

    def custom_b64decode(input):
        return base64.b64decode(input.encode('utf-8')).decode('utf-8')

The function above will base-64 decode the value returned from the Secrets
Manager and finally give us the ``secret`` value we need.

'''

import os
import os.path
import getpass

######################
## USEFUL CONSTANTS ##
######################

this_dir = os.path.abspath(os.path.dirname(__file__))
default_permissions_file = os.path.abspath(
    os.path.join(this_dir,
                 'default-permissions-model.json')
)
currentuser = getpass.getuser()

ENVPREFIX = 'AUTHNZERVER'


###############
## MAIN CONF ##
###############

CONF = {
    'authdb':{
        'env':'%s_AUTHDB' % ENVPREFIX,
        'cmdline':'authdb',
        'type':str,
        'default':None,
        'help':('An SQLAlchemy database URL to indicate where '
                'the local authentication DB is. '
                'This should be in the form discussed at: '
                'https://docs.sqlalchemy.org/en/latest'
                '/core/engines.html#database-urls'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'basedir':{
        'env':'%s_BASEDIR' % ENVPREFIX,
        'cmdline':'basedir',
        'type':str,
        'default':os.getcwd(),
        'help':('The base directory containing secret files and the auth DB.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'cachedir':{
        'env':'%s_CACHEDIR' % ENVPREFIX,
        'cmdline':'cachedir',
        'type':str,
        'default':'/tmp/authnzerver-cache',
        'help':('Path to the cache directory to be used.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'debugmode':{
        'env':'%s_DEBUGMODE' % ENVPREFIX,
        'cmdline':'debugmode',
        'type':int,
        'default':False,
        'help':('If 1, will enable an '
                '/echo endpoint for debugging purposes.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'listen':{
        'env':'%s_LISTEN' % ENVPREFIX,
        'cmdline':'listen',
        'type':str,
        'default':'127.0.0.1',
        'help':('Bind to this address and serve content.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'permissions':{
        'env':'%s_PERMISSIONS' % ENVPREFIX,
        'cmdline':'permissions',
        'type':str,
        'default':default_permissions_file,
        'help':('The JSON file containing the permissions '
                'model the server will enforce.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'port':{
        'env':'%s_PORT' % ENVPREFIX,
        'cmdline':'port',
        'type':int,
        'default':13431,
        'help':('Run the server on this TCP port.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'secret':{
        'env':'%s_SECRET' % ENVPREFIX,
        'cmdline':'secret',
        'type':str,
        'default':None,
        'help':('The shared secret key used to secure '
                'communications between authnzerver and any frontend servers.'),
        'readable_from_file':'string',
        'postprocess_value':None,
    },
    'piisalt':{
        'env':'%s_PIISALT' % ENVPREFIX,
        'cmdline':'piisalt',
        'type':str,
        'default':None,
        'help':('A random value used as a salt when SHA256 hashing personally '
                'identifiable information (PII), such as user IDs and '
                'session tokens, etc. for authnzerver logs.'),
        'readable_from_file':'string',
        'postprocess_value':None,
    },
    'sessionexpiry':{
        'env':'%s_SESSIONEXPIRY' % ENVPREFIX,
        'cmdline':'sessionexpiry',
        'type':int,
        'default':30,
        'help':('This sets the session-expiry time in days.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'workers':{
        'env':'%s_WORKERS' % ENVPREFIX,
        'cmdline':'workers',
        'type':int,
        'default':4,
        'help':('The number of background workers '
                'to use when processing requests.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'emailserver':{
        'env':'%s_EMAILSERVER' % ENVPREFIX,
        'cmdline':'emailserver',
        'type':str,
        'default':'localhost',
        'help':('The address of the email server to use.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'emailport':{
        'env':'%s_EMAILPORT' % ENVPREFIX,
        'cmdline':'emailport',
        'type':int,
        'default':25,
        'help':('The SMTP port of the email server to use.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'emailuser':{
        'env':'%s_EMAILUSER' % ENVPREFIX,
        'cmdline':'emailuser',
        'type':str,
        'default':currentuser,
        'help':('The username to use for login to the email server.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'emailpass':{
        'env':'%s_EMAILPASS' % ENVPREFIX,
        'cmdline':'emailpass',
        'type':str,
        'default':'',
        'help':('The password to use for login to the email server.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'emailsender':{
        'env':'%s_EMAILSENDER' % ENVPREFIX,
        'cmdline':'emailsender',
        'type':str,
        'default':'Authnzerver <authnzerver@localhost>',
        'help':('The account name and email address that the '
                'authnzerver will send from.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'tls_cert_file':{
        'env':'%s_TLSCERT_FILE' % ENVPREFIX,
        'cmdline':'tlscertfile',
        'type':str,
        'default':'',
        'help':('The TLS certificate to use. If this is provided along '
                'with the certificate key in the --tlscertkey option, '
                'the server will start in TLS-enabled mode.'),
        'readable_from_file':False,
        'postprocess_value':None,
    },
    'tls_cert_key':{
        'env':'%s_TLSCERT_KEY' % ENVPREFIX,
        'cmdline':'tlscertkey',
        'type':str,
        'default':'',
        'help':("The TLS certificate's key to use. If this is provided along "
                "with the certificate in the --tlscertfile option, "
                "the server will start in TLS-enabled mode."),
        'readable_from_file':False,
        'postprocess_value':None,
    },
}
